File: infrastructure/ts-sdk/src/graceful.ts
/**
 * Graceful shutdown and health probe support for TypeScript services.
 */

import { Server } from 'http';

export interface ComponentHealth {
  name: string;
  connected: boolean;
  latencyMs: number;
  error?: string;
}

export type HealthCheckFn = () => Promise<boolean>;

export class HealthRegistry {
  private checks: Map<string, HealthCheckFn> = new Map();
  private started: number = Date.now();
  private ready = true;
  private serviceName: string;

  constructor(serviceName: string) {
    this.serviceName = serviceName;
  }

  register(name: string, checkFn: HealthCheckFn): void {
    this.checks.set(name, checkFn);
  }

  setReady(ready: boolean): void {
    this.ready = ready;
  }

  get isReady(): boolean {
    return this.ready;
  }

  async checkAll(): Promise<Record<string, unknown>> {
    const results: Record<string, ComponentHealth> = {};
    let allHealthy = true;

    for (const [name, checkFn] of this.checks) {
      const start = Date.now();
      try {
        const ok = await Promise.race([
          checkFn(),
          new Promise<boolean>((_, reject) =>
            setTimeout(() => reject(new Error('timeout')), 5000),
          ),
        ]);
        results[name] = {
          name,
          connected: ok,
          latencyMs: Date.now() - start,
        };
        if (!ok) allHealthy = false;
      } catch (error: any) {
        allHealthy = false;
        results[name] = {
          name,
          connected: false,
          latencyMs: Date.now() - start,
          error: error.message,
        };
      }
    }

    return {
      status: allHealthy ? 'healthy' : 'degraded',
      service: this.serviceName,
      uptimeSeconds: (Date.now() - this.started) / 1000,
      components: results,
    };
  }

  healthResponse(): Record<string, unknown> {
    return {
      status: this.ready ? 'healthy' : 'degraded',
      service: this.serviceName,
      uptimeSeconds: (Date.now() - this.started) / 1000,
    };
  }

  livenessResponse(): Record<string, unknown> {
    return { alive: true, service: this.serviceName };
  }

  readinessResponse(): Record<string, unknown> {
    return { ready: this.ready, service: this.serviceName };
  }
}

export class GracefulShutdown {
  private cleanupHooks: Array<() => Promise<void>> = [];
  private shuttingDown = false;
  private serviceName: string;

  constructor(serviceName: string) {
    this.serviceName = serviceName;
  }

  addCleanup(fn: () => Promise<void>): void {
    this.cleanupHooks.push(fn);
  }

  setupSignals(server?: Server): void {
    const handler = async (signal: string) => {
      if (this.shuttingDown) {
        console.error(`[${this.serviceName}] Force shutdown`);
        process.exit(1);
      }

      this.shuttingDown = true;
      console.log(`[${this.serviceName}] ${signal} received, shutting down gracefully...`);

      if (server) {
        server.close(() => {
          console.log(`[${this.serviceName}] HTTP server closed`);
        });
      }

      await this.cleanup();
      process.exit(0);
    };

    process.on('SIGINT', () => handler('SIGINT'));
    process.on('SIGTERM', () => handler('SIGTERM'));
  }

  async cleanup(): Promise<void> {
    console.log(`[${this.serviceName}] Running cleanup hooks...`);
    for (const hook of this.cleanupHooks.reverse()) {
      try {
        await Promise.race([
          hook(),
          new Promise<void>((_, reject) =>
            setTimeout(() => reject(new Error('cleanup timeout')), 10000),
          ),
        ]);
      } catch (error: any) {
        console.error(`[${this.serviceName}] Cleanup hook failed: ${error.message}`);
      }
    }
    console.log(`[${this.serviceName}] Graceful shutdown complete`);
  }

  get isShuttingDown(): boolean {
    return this.shuttingDown;
  }
}

File: infrastructure/python-sdk/infra_sdk/graceful.py
"""Graceful shutdown and health probe support for Python services."""

import asyncio
import logging
import os
import signal
import time
from dataclasses import dataclass, field
from typing import Any, Callable, Dict, List, Optional

logger = logging.getLogger(__name__)


@dataclass
class ComponentHealth:
    name: str
    connected: bool
    latency_ms: float = 0.0
    error: str = ""


class HealthRegistry:
    """Tracks health of all registered components."""

    def __init__(self, service_name: str):
        self.service_name = service_name
        self.started = time.time()
        self._checks: Dict[str, Callable] = {}
        self._ready = True

    def register(self, name: str, check_fn: Callable) -> None:
        self._checks[name] = check_fn

    def set_ready(self, ready: bool) -> None:
        self._ready = ready

    @property
    def is_ready(self) -> bool:
        return self._ready

    async def check_all(self) -> Dict[str, Any]:
        results = {}
        all_healthy = True

        for name, check_fn in self._checks.items():
            start = time.time()
            try:
                if asyncio.iscoroutinefunction(check_fn):
                    await asyncio.wait_for(check_fn(), timeout=5.0)
                else:
                    check_fn()
                latency_ms = (time.time() - start) * 1000
                results[name] = ComponentHealth(
                    name=name, connected=True, latency_ms=latency_ms
                )
            except Exception as e:
                latency_ms = (time.time() - start) * 1000
                all_healthy = False
                results[name] = ComponentHealth(
                    name=name, connected=False, latency_ms=latency_ms, error=str(e)
                )

        return {
            "status": "healthy" if all_healthy else "degraded",
            "service": self.service_name,
            "uptime_seconds": time.time() - self.started,
            "components": {
                name: {
                    "connected": h.connected,
                    "latency_ms": round(h.latency_ms, 2),
                    "error": h.error,
                }
                for name, h in results.items()
            },
        }

    def health_response(self) -> Dict[str, Any]:
        """Synchronous health response for simple checks."""
        return {
            "status": "healthy" if self._ready else "degraded",
            "service": self.service_name,
            "uptime_seconds": round(time.time() - self.started, 2),
        }

    def liveness_response(self) -> Dict[str, Any]:
        return {"alive": True, "service": self.service_name}

    def readiness_response(self) -> Dict[str, Any]:
        return {"ready": self._ready, "service": self.service_name}


class GracefulShutdown:
    """Manages graceful shutdown with signal handling and cleanup hooks."""

    def __init__(self, service_name: str):
        self.service_name = service_name
        self._cleanup_hooks: List[Callable] = []
        self._shutdown_event = asyncio.Event()
        self._shutting_down = False

    def add_cleanup(self, fn: Callable) -> None:
        self._cleanup_hooks.append(fn)

    def setup_signals(self, loop: Optional[asyncio.AbstractEventLoop] = None) -> None:
        loop = loop or asyncio.get_event_loop()

        for sig in (signal.SIGINT, signal.SIGTERM):
            loop.add_signal_handler(sig, lambda s=sig: self._handle_signal(s))

        logger.info("Signal handlers registered for %s", self.service_name)

    def _handle_signal(self, sig: signal.Signals) -> None:
        if self._shutting_down:
            logger.warning("Force shutdown requested for %s", self.service_name)
            os._exit(1)

        self._shutting_down = True
        logger.info(
            "Shutdown signal %s received for %s", sig.name, self.service_name
        )
        self._shutdown_event.set()

    async def wait_for_shutdown(self) -> None:
        await self._shutdown_event.wait()

    async def cleanup(self) -> None:
        logger.info("Running cleanup hooks for %s", self.service_name)
        for hook in reversed(self._cleanup_hooks):
            try:
                if asyncio.iscoroutinefunction(hook):
                    await asyncio.wait_for(hook(), timeout=10.0)
                else:
                    hook()
            except Exception as e:
                logger.error("Cleanup hook failed: %s", e)
        logger.info("Graceful shutdown complete for %s", self.service_name)

    @property
    def is_shutting_down(self) -> bool:
        return self._shutting_down
